### PR TITLE
Allow to route comms over bokeh's protocol

### PR DESCRIPTION
**Pull request by mattpap**
This is a set of necessary changes to allow https://github.com/bokeh/ipywidgets_bokeh to work. This adds `MessageSent` document-level event, which allows bi-directional routing of arbitrary objects (with resolving of references, etc.). Originally this was primarily intended for routing comms, but as a generic mechanism, it can route e.g. "bokeh" events. I took this opportunity and reimplemented them using `MessageSent`. This way we don't need protocol-level support for them and all the unnecessary accompanying logic, and we can use the existing machinery to e.g. allow passing object references directly, perhaps allow binary data transfer, etc.

A separate, but related change is to maintain current document during `apply_json_patch()`. This allows objects implicitly associated with a document to communicate over the correct channel, without a need to keep an explicit registry (which in general is not possible).

### CODE REVIEW

**Comment by bryevdv** [comment]
Well, the release notes/migration guide will need to be updated to list all the API removals and changes. 

I think at this point we need to have a high level "map"  that lists what is handled directly at the protocol level, and what is handled as some "subprotocol" inside another message type. That could be a comment, or actual docs in a docstring for the protocol module, etc. 

Do you intend to move other existing messages under `MessageSentEvent`? What is the overall plan and end goal here?

**Comment by mattpap** [comment]
> Well, the release notes/migration guide will need to be updated to list all the API removals and changes.

From users' perspective nothing changes. Though changes to the protocol should be documented.

> Do you intend to move other existing messages under MessageSentEvent? What is the overall plan and end goal here?

I don't intend any further changes, at least to the document events themselves. The current situation, where we have strongly typed events for the most part and this new one for communicating any data, is fine. However, I may unify handling of references, binary protocol, etc. in another PR.

**Comment by mattpap** [comment]
@philippjfr, `jupyter_embed` is up, so now you just need to compile the extension locally and should be able run the example.

**Comment by bryevdv** [comment]
>  Though changes to the protocol should be documented.

Right, we are are on the same page, I was only suggesting that the release notes are the right place for that to happens

Everything else sounds great.

File: bokeh/document/events.py [hunk 1]
```diff
@@ -44,6 +44,7 @@
     'SessionCallbackRemoved',
     'TitleChangedEvent',
     'TitleChangedEvent',
```
Comment (xavArtley) [inline bokeh/document/events.py]: Hello TitleChangedEvent seems duplicate

**Comment by mattpap** [comment]
@philippjfr, did you have a chance to test this out?

**Comment by philippjfr** [comment]
I've been somewhat swamped I'll give it a try this evening.

File: bokeh/document/events.py (hunk 1 above)
Comment (mattpap) [reply bokeh/document/events.py]: Thanks. Fixed.

**Comment by philippjfr** [comment]
Working well for me! Just tried it out using your example and some other ipywidgets based examples as well as using it from Panel. All seems to work well.

**Pull request merged**